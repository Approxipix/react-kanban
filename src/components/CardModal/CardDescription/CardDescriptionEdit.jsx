import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { editCardDesc } from '../../../redux/cardReducer/actions';
import { Actions, CancelButton, SubmitButton, TextArea } from '../../BaseComponent';
import ClickOutside from '../../ClickOutside';
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";

const textareaLineHeight = 16;

class CardDescriptionEdit extends Component {
  constructor(props) {
    super(props);
    this.state = {
      description: props.cardDescription || '',
      rows: 2,
      minRows: 2,
    };
  }

  componentDidMount() {
    this.setState({
      rows: ~~(this.textarea.scrollHeight / textareaLineHeight)
    })
  }

  handleKeyDown = (e) => {
    if (e.keyCode === 27) {
      this.props.toggleEditDescription()
    }
  };

  handleChange = (e) => {
    const { minRows } = this.state;
    const previousRows = e.target.rows;
    e.target.rows = minRows; // reset number of rows in textarea
    e.target.style.overflow = 'hidden'; //overflow because scrollbar affects e.target.scrollHeight
    const currentRows = ~~(e.target.scrollHeight / textareaLineHeight);
    if (currentRows === previousRows) {
      e.target.rows = currentRows;
    }
    this.setState({
      description: e.target.value,
      rows: currentRows,
    });
  };

  handleSubmit = (e) => {
    e.preventDefault();
    const { description } = this.state;
    const { cardId } = this.props;
    this.props.actions.editCardDesc({
      cardDescription: description,
      cardId: cardId,
    });
    this.props.toggleEditDescription()
  };

  render() {
    const { description, rows } = this.state;
    return (
      <ClickOutside toggleOpened={this.props.toggleEditDescription}>
        <form>
          <TextArea
            type="text"
            size=".875rem;"
            padding=".5rem"
            value={description}
            rows={rows}
            placeholder="Edit card description"
            ref={(e) => { this.textarea = e }}
            onKeyDown={this.handleKeyDown}
            onChange={(e) => this.handleChange(e)}
            onBlur={() => this.props.toggleEditDescription()}
            autoFocus
            spellCheck={false}
          />
          <Actions>
            <SubmitButton onMouseDown={(e) => this.handleSubmit(e)} >
              Save
            </SubmitButton>
            <CancelButton onClick={() => this.props.toggleEditDescription()}>
              <FontAwesomeIcon icon="times" />
            </CancelButton>
          </Actions>
        </form>
      </ClickOutside>
    );
  };
}

function mapDispatchToProps(dispatch) {
  return {
    actions: bindActionCreators({
      editCardDesc: editCardDesc,
    }, dispatch)
  };
}

export default connect(null, mapDispatchToProps)(CardDescriptionEdit);
