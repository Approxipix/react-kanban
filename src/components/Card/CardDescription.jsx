import React, { Component } from 'react';
import styled from 'styled-components';
import CardDescriptionEdit from './CardDescriptionEdit';

const Desc = styled.h4`
  margin-bottom: .5rem;
  padding: .3rem .5rem;
  font-size: 1.5rem;
  font-weight: 600;
`;

const AddDescription = styled.div``;

class CardDescription extends Component {
  constructor(props) {
    super(props);
    this.state = {
      isDescEditing: false,
    }
  }

  toggleIsDescEditing = () => {
    this.setState({
      isDescEditing: !this.state.isDescEditing
    })
  };

  render() {
    const { cardId, cardDescription } = this.props;
    const { isDescEditing } = this.state;

    if (isDescEditing) {
      return (
        <CardDescriptionEdit
          cardId={cardId}
          cardDescription={cardDescription}
          toggleIsDescEditing={this.toggleIsDescEditing}
        />
      )
    }
    if (!cardDescription) {
      return (
        <AddDescription onClick={() => this.toggleIsDescEditing()}>
          Add a more detailed descriptionâ€¦
        </AddDescription>
      )
    }
    return (
      <Desc onClick={() => this.toggleIsDescEditing()}>
        {cardDescription}
      </Desc>
    )
  }
}


export default CardDescription;
