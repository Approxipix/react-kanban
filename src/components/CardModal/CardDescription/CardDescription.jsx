import React, { Component } from 'react';
import CardDescriptionEdit from './CardDescriptionEdit';
import styled from 'styled-components';

const Description = styled.p`
  font-size: .875rem;
  word-break: break-word;
  white-space: pre-line;
  cursor: pointer;
`;

const AddDescription = styled.a`
  display: flex;
  padding: .5rem .75rem;
  width: 100%;
  min-height: 4rem;
  font-size: .875rem;
  text-align: left;
  background-color: rgba(9, 30, 66, .08);
  border-radius: 3px;
  transition: background-color .2s ease-in;
  cursor: pointer;
  &:hover {
    background-color: rgba(9,30,66,.13);
  }
`;

const Button = styled.button`
  position: absolute;
  top: 0;
  right: 0;
  padding: .5rem 1rem;
  background-color: rgba(9, 30, 66, .08);
  border-radius: 3px;
  transition: background-color .2s ease-in;
  &:hover {
    background-color: rgba(9,30,66,.13);
  }
`;

class CardDescription extends Component {
  constructor(props) {
    super(props);
    this.state = {
      editDescription: false,
    }
  }

  toggleEditDescription = () => {
    this.setState({
      editDescription: !this.state.editDescription
    })
  };

  render() {
    const { cardId, cardDescription } = this.props;
    const { editDescription } = this.state;
    if (!cardDescription && !editDescription) {
      return (
        <AddDescription onClick={() => this.toggleEditDescription()}>
          Add a more detailed description...
        </AddDescription>
      )
    }
    return (
      !editDescription ? (
        <Description>
          {cardDescription}
          <Button onClick={() => this.toggleEditDescription()}>
            Edit
          </Button>
        </Description>
      ) : (
        <CardDescriptionEdit
          cardId={cardId}
          cardDescription={cardDescription}
          toggleEditDescription={this.toggleEditDescription}
        />
      )
    )
  }
}

export default CardDescription;
