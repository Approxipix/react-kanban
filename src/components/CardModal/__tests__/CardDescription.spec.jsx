import React from 'react';
import { shallow } from 'enzyme';
import toJson from "enzyme-to-json";
import CardDescription from '../CardDescription/CardDescription';

describe('<CardDescription>', () => {
  let props;
  let state;
  let component;

  describe('when cardDescription prop is exist', () => {
    beforeEach(() => {
      props = { cardDescription: 'Card Description' };
      state = { editDescription: false };

      component = shallow(<CardDescription {...props} />);
      component.setState({...state});
    });

    it('should render Edit button if editDescription is false', () => {
      expect(toJson(component)).toMatchSnapshot();
    });

  });

  describe('when cardDescription prop does not exist', () => {
    beforeEach(() => {
      props = { cardDescription: '' };
      state = { editDescription: false };

      component = shallow(<CardDescription {...props} />);
      component.setState({...state});
    });

    it('should render AddDescription if editDescription is false', () => {
      expect(toJson(component)).toMatchSnapshot();
    });

    it("should call toggleEditDescription after click on AddDescription", () => {
      const spy = jest.spyOn(component.instance(), "toggleEditDescription");

      component.find('AddDescription').simulate('click');

      expect(spy).toHaveBeenCalledTimes(1);
    });

    it('should render CardDescriptionEdit if editDescription is true', () => {
      component.setState({ editDescription: true });
      expect(toJson(component)).toMatchSnapshot();
    });
  });

  it('checks if toggleEditDescription sets editDescription correctly', () => {
    component.setState({ editDescription: false });

    component.instance().toggleEditDescription();
    expect(component.instance().state.editDescription).toEqual(true);

    component.instance().toggleEditDescription();
    expect(component.instance().state.editDescription).toEqual(false);
  });
});
