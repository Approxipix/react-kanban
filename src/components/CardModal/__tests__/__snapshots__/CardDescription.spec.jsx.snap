// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CardDescription> when cardDescription prop does not exist should render AddDescription if editDescription is false 1`] = `
<AddDescription
  onClick={[Function]}
>
  Add a more detailed description...
</AddDescription>
`;

exports[`<CardDescription> when cardDescription prop does not exist should render CardDescriptionEdit if editDescription is true 1`] = `
<Connect(CardDescriptionEdit)
  cardDescription=""
  cardId=""
  toggleEditDescription={[Function]}
/>
`;

exports[`<CardDescription> when cardDescription prop is exist should render Edit button if editDescription is false 1`] = `
<styled.p>
  Card Description
  <styled.button
    onClick={[Function]}
  >
    Edit
  </styled.button>
</styled.p>
`;
